<a href="https://colab.research.google.com/github/darpan02-cypher/Knowledge-Data-and-Discovery/blob/main/PatternMining_encoding02_HW2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

The mlxtend library is a collection of useful tools and extensions for the Python data science and machine learning libraries (like pandas, scikit-learn, and matplotlib). It includes various functionalities for tasks such as:

Data preprocessing and transformation
Feature extraction and selection
Model evaluation and visualization
Statistical tests
Association rule learning (like the Apriori algorithm)

In [1]:
pip install mlxtend

In [2]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder

In [3]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [4]:
%cd /content/drive/My Drive/Colab Notebooks

/content/drive/My Drive/Colab Notebooks


In [5]:
#Reading data into a dataframe
titanic = pd.read_csv('/content/drive/MyDrive/KDD/Titanic.csv')

In [6]:
titanic.head(10)

Class   Sex    Age Survived
0   3rd  Male  Child       No
1   3rd  Male  Child       No
2   3rd  Male  Child       No
3   3rd  Male  Child       No
4   3rd  Male  Child       No
5   3rd  Male  Child       No
6   3rd  Male  Child       No
7   3rd  Male  Child       No
8   3rd  Male  Child       No
9   3rd  Male  Child       No

In [7]:
# Map the 'Survived' variable to Boolean values
titanic['Survived'] = titanic['Survived'].map({'Yes': 1, 'No': 0})

In [8]:
#Transform into Pandas DataFrame
# One-hot encode categorical variables: 'Class', 'Sex', and 'Age'
categorical_columns = ['Class', 'Sex', 'Age']
encoder = OneHotEncoder(sparse_output=False, drop=None)
encoded_columns = encoder.fit_transform(titanic[categorical_columns])
encoded_df = pd.DataFrame(encoded_columns, columns=encoder.get_feature_names_out(categorical_columns))
titanic = titanic.drop(categorical_columns, axis=1)
titanic = pd.concat([titanic, encoded_df], axis=1)

In [9]:
titanic

Survived  Class_1st  Class_2nd  Class_3rd  Class_Crew  Sex_Female  \
0            0        0.0        0.0        1.0         0.0         0.0   
1            0        0.0        0.0        1.0         0.0         0.0   
2            0        0.0        0.0        1.0         0.0         0.0   
3            0        0.0        0.0        1.0         0.0         0.0   
4            0        0.0        0.0        1.0         0.0         0.0   
...        ...        ...        ...        ...         ...         ...   
2196         1        0.0        0.0        0.0         1.0         1.0   
2197         1        0.0        0.0        0.0         1.0         1.0   
2198         1        0.0        0.0        0.0         1.0         1.0   
2199         1        0.0        0.0        0.0         1.0         1.0   
2200         1        0.0        0.0        0.0         1.0         1.0   

      Sex_Male  Age_Adult  Age_Child  
0          1.0        0.0        1.0  
1          1.0        0.0        1.0  
2          1.0        0.0        1.0  
3          1.0        0.0        1.0  
4          1.0        0.0        1.0  
...        ...        ...        ...  
2196       0.0        1.0        0.0  
2197       0.0        1.0        0.0  
2198       0.0        1.0        0.0  
2199       0.0        1.0        0.0  
2200       0.0        1.0        0.0  

[2201 rows x 9 columns]

In [10]:
#Asked 35 frequent itemsets with the support level 0.1

from mlxtend.frequent_patterns import fpgrowth, association_rules
frequent_itemsets = fpgrowth(titanic, min_support=0.1, use_colnames=True)

/usr/local/lib/python3.12/dist-packages/mlxtend/frequent_patterns/fpcommon.py:161: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: Dep

In [11]:
print(frequent_itemsets)

     support                           itemsets
0   0.786461                         (Sex_Male)
1   0.320763                        (Class_3rd)
2   0.213539                       (Sex_Female)
3   0.950477                        (Age_Adult)
4   0.147660                        (Class_1st)
5   0.129487                        (Class_2nd)
6   0.402090                       (Class_Crew)
7   0.323035                         (Survived)
8   0.757383              (Age_Adult, Sex_Male)
9   0.231713              (Sex_Male, Class_3rd)
10  0.284871             (Age_Adult, Class_3rd)
11  0.209905   (Age_Adult, Sex_Male, Class_3rd)
12  0.193094            (Sex_Female, Age_Adult)
13  0.156293             (Sex_Female, Survived)
14  0.143571  (Sex_Female, Age_Adult, Survived)
15  0.144934             (Age_Adult, Class_1st)
16  0.118582             (Class_2nd, Age_Adult)
17  0.402090            (Class_Crew, Age_Adult)
18  0.391640             (Class_Crew, Sex_Male)
19  0.391640  (Class_Crew, Age_Adult, Se

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [12]:
top_35_itemsets = frequent_itemsets.head(35)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [13]:
#Correct initial result of the associate rules with the confidence level 0.7
association_rules_df = association_rules(frequent_itemsets, metric="confidence", min_threshold=0.7)
association_rules_df

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents            consequents  antecedent support  \
0               (Age_Adult)             (Sex_Male)            0.950477   
1                (Sex_Male)            (Age_Adult)            0.786461   
2               (Class_3rd)             (Sex_Male)            0.320763   
3               (Class_3rd)            (Age_Adult)            0.320763   
4    (Age_Adult, Class_3rd)             (Sex_Male)            0.284871   
5     (Sex_Male, Class_3rd)            (Age_Adult)            0.231713   
6              (Sex_Female)            (Age_Adult)            0.213539   
7              (Sex_Female)             (Survived)            0.213539   
8   (Sex_Female, Age_Adult)             (Survived)            0.193094   
9    (Sex_Female, Survived)            (Age_Adult)            0.156293   
10              (Class_1st)            (Age_Adult)            0.147660   
11              (Class_2nd)            (Age_Adult)            0.129487   
12             (Class_Crew)            (Age_Adult)            0.402090   
13             (Class_Crew)             (Sex_Male)            0.402090   
14  (Class_Crew, Age_Adult)             (Sex_Male)            0.402090   
15   (Class_Crew, Sex_Male)            (Age_Adult)            0.391640   
16             (Class_Crew)  (Age_Adult, Sex_Male)            0.402090   
17               (Survived)            (Age_Adult)            0.323035   
18     (Sex_Male, Survived)            (Age_Adult)            0.166742   

    consequent support   support  confidence      lift  representativity  \
0             0.786461  0.757383    0.796845  1.013204               1.0   
1             0.950477  0.757383    0.963027  1.013204               1.0   
2             0.786461  0.231713    0.722380  0.918520               1.0   
3             0.950477  0.284871    0.888102  0.934375               1.0   
4             0.786461  0.209905    0.736842  0.936909               1.0   
5             0.950477  0.209905    0.905882  0.953082               1.0   
6             0.950477  0.193094    0.904255  0.951370               1.0   
7             0.323035  0.156293    0.731915  2.265745               1.0   
8             0.323035  0.143571    0.743529  2.301699               1.0   
9             0.950477  0.143571    0.918605  0.966467               1.0   
10            0.950477  0.144934    0.981538  1.032680               1.0   
11            0.950477  0.118582    0.915789  0.963505               1.0   
12            0.950477  0.402090    1.000000  1.052103               1.0   
13            0.786461  0.391640    0.974011  1.238474               1.0   
14            0.786461  0.391640    0.974011  1.238474               1.0   
15            0.950477  0.391640    1.000000  1.052103               1.0   
16            0.757383  0.391640    0.974011  1.286022               1.0   
17            0.950477  0.297138    0.919831  0.967757               1.0   
18            0.950477  0.153567    0.920981  0.968967               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
0   0.009870    1.051116       0.263149  0.773191   0.048630    0.879936  
1   0.009870    1.339441       0.061028  0.773191   0.253420    0.879936  
2  -0.020555    0.769177      -0.115514  0.264660  -0.300090    0.508503  
3  -0.020008    0.442572      -0.093712  0.288807  -1.259519    0.593908  
4  -0.014135    0.811449      -0.086060  0.243671  -0.232363    0.501870  
5  -0.010333    0.526181      -0.060217  0.215888  -0.900486    0.563362  
6  -0.009870    0.517240      -0.061028  0.198877  -0.933340    0.553705  
7   0.087312    2.525187       0.710327  0.410992   0.603990    0.607870  
8   0.081195    2.639542       0.700873  0.385366   0.621146    0.593987  
9  -0.004981    0.608425      -0.039500  0.149057  -0.643589    0.534828  
10  0.004587    2.682493       0.037128  0.152050   0.627212    0.567012  
11 -0.004492    0.588085      -0.041697  0.123346  -0.700435    0.520275  
12  0.019913         inf       0.082827

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [14]:
#consequents only being a value of the Survived variable
association_rules_df = association_rules_df[association_rules_df['consequents'].apply(lambda x: len(x) == 1 and 'Survived' in x)]
association_rules_df


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents consequents  antecedent support  \
7             (Sex_Female)  (Survived)            0.213539   
8  (Sex_Female, Age_Adult)  (Survived)            0.193094   

   consequent support   support  confidence      lift  representativity  \
7            0.323035  0.156293    0.731915  2.265745               1.0   
8            0.323035  0.143571    0.743529  2.301699               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
7  0.087312    2.525187       0.710327  0.410992   0.603990    0.607870  
8  0.081195    2.639542       0.700873  0.385366   0.621146    0.593987

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


In [15]:
# Now display valid association rules with the relevant information
association_rules_df[['antecedents', 'consequents', 'support', 'confidence']]

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents consequents   support  confidence
7             (Sex_Female)  (Survived)  0.156293    0.731915
8  (Sex_Female, Age_Adult)  (Survived)  0.143571    0.743529

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [16]:
association_rules_df.sort_values("confidence",ascending=False)

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

antecedents consequents  antecedent support  \
8  (Sex_Female, Age_Adult)  (Survived)            0.193094   
7             (Sex_Female)  (Survived)            0.213539   

   consequent support   support  confidence      lift  representativity  \
8            0.323035  0.143571    0.743529  2.301699               1.0   
7            0.323035  0.156293    0.731915  2.265745               1.0   

   leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
8  0.081195    2.639542       0.700873  0.385366   0.621146    0.593987  
7  0.087312    2.525187       0.710327  0.410992   0.603990    0.607870

/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:151: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)


Summary of Titanic Survival Analysis using Association Rule Mining

This notebook aimed to discover interesting relationships between passenger characteristics and survival on the Titanic using association rule mining.

The process involved:

Data Loading and Preparation: The Titanic dataset was loaded and preprocessed. This included converting the 'Survived' column to numerical values (1 for Yes, 0 for No) and applying one-hot encoding to the categorical features ('Class', 'Sex', and 'Age') to prepare the data for the association rule mining algorithm.

Frequent Itemset Generation: The FPGrowth algorithm was used to identify frequent itemsets (combinations of characteristics) that appear together in the dataset with a minimum support of 10%.
Association Rule Mining: Association rules were generated from the frequent itemsets with a minimum confidence threshold of 70%. These rules highlight potential relationships between different characteristics and survival.

Analysis of Survival Rules: The generated rules were filtered to focus specifically on those where 'Survived' was the consequent. This allowed for the identification of characteristics or combinations of characteristics that strongly predict survival.
Key Findings (Based on the filtered rules with Survived as consequent):

The analysis revealed strong associations between certain passenger characteristics and survival:

The rule {Sex_Female} -> {Survived} shows that being female is strongly associated with survival, with a confidence of approximately 73.2%.
The rule {Sex_Female, Age_Adult} -> {Survived} indicates that being an adult female is also strongly associated with survival, with a confidence of approximately 74.4%.
These findings suggest that gender, and to some extent age, played a significant role in survival rates on the Titanic. The higher confidence values for rules involving 'Sex_Female' highlight the well-known "women and children first" protocol that was followed during the disaster.